Enable writing of 3-D netCDF data cubes from greenspline and grdinterpolate

PaulWessel opened the pull request:
**Description of proposed changes**

This PR completes the _gmt_write_nc_cube_ coding to enable the writing of 3-D layers (with layers an unlimited dimension).  It is WIP because I want to complete what GMT_Read_Data with **CONTAINER_ONLY** should do for cubes, and update gspline_4.sh to no longer write lots of 2-D grids.

Related question whose solution can be separate PR is:

- Do we add **-Q** for "cube" options to **grd2xyz** and **xyz2grd** so they can handle 4-D data or do we clone and write **cube2xyzw**/**xyzw2cube**?  My vote would be for -Q since adding two new modules for this seems overkill. But we do want the ability to dump and load cubes.
- What about **grdinfo** for 3-D cubes? We currently have no way to determine the equivalent info for cubes that we create.  I see **-Q** is available there too.

PaulWessel commented:
Alternatively, no **-Q** is needed for **grdinfo** and **grd2xyz** since we could presumably check if a given netCDF file is a grid or cube, and then take action.  As for **xyz2grd** one has to specify -Rw/e/s/n -Idx/dy, but if a cube this needs to be -Rw/e/s/n/z0/z1 -Idx/dxy/dz so perhaps there is no need for **-Q** there either.  What do you think, @joa-quim ?

joa-quim commented:
> grd2xyz and xyz2grd so they can handle 4-D data

What is the interest of this? Why care to convert a 4-D to *xyzt* and back?
A more useful tool would a stacker, which would read a pack of grids and stack them into a 3-D nc file.

PaulWessel commented:
`grdinterpolate grid1 grid2 ... -Zlevels -Gcube.nc `will do the stacking (in next commit).
We cared enough to convert grd <--> xyz so I guess it is not that different to do cube <--> xyzw ?

joa-quim commented:
> We cared enough to convert grd <--> xyz 

>20 years ago when people still wrote grids manually. I fail to see why spending time coding a 4D analog

PaulWessel commented:
OK, we can wait to see if there is any demand. What about grdinfo for cubes?

PaulWessel commented on src/gmt_nc.c:
```diff
@@ -392,10 +396,35 @@ GMT_LOCAL bool gmtnc_not_obviously_polar (double *se) {
+
+	/* Depending on row_order, write y-coordinate array bottom-to-top or top-to-bottom */
+	if (HH->row_order == k_nc_start_south) {
+		for (row = 0; row < header->n_rows; row++) xy[row] = (double) gmt_M_col_to_x (GMT, row, header->wesn[YLO], header->wesn[YHI], header->inc[GMT_Y], 0.5 * header->registration, header->n_rows);
```
OK, done in my workspace - not ready to commit yet.

joa-quim commented:
``grdinfo`` for cubes seems useful.

PaulWessel commented:
Very close to finishing this PR (dealing with consistency and detecting z-axis being absolute time) but running into some wider issues related to netCDF COARDS attributes that are getting worse by now dealing with 4-D data = (_x,y,z,v_):

1. When writing netCDF grids (and now cubes) we have **no ability** to name the data variable (it is always called "z" for grids).  In contrast, the x and y variables get proper names if geographic data (lon, lat), else they are always called x,y (or time if appropriate).  I think these names for the dimensions are COARDS specifications.
2. The netCDF  _attributes_ for x,y also gets suitable names if geo (e.g., longitude).  We can set the z unit via grdedit -D (or -D in xyz2grd) but what that does it to set the **name** attribute (not the unit attribute).  The variable name remains z.
3. Our grid header only has one text string per dimension, e.g. header->x_unit and it may typically say "longitude" which is not exactly a "unit".  However, when a netCDF file is created, we intervene and split this into a name (longitude), a unit (degrees_east), and even add two more attributes: standard name (longitude), and Axis (X).  For z there is no such treatment because the x/y can be done correctly if we know they are geographic.  So header->z_unit is the only thing we have and write.
4. GMT does not care if the grid "z" variable is called "z" or anything else (e.g., "gravity") as it will find the first 2-D variable in the file.  If there are _several_ 2-D variables then you must specify its name as part of the file name ("collection.grd?gravity").  So reading is fine, it is just that we cannot (in GMT) specify a sensible variable name for z when we create files.
5. For data cubes, the "z" is instead just the 3rd dimension, and for most cases it will either be a spatial one (e.g., depth, height level) or temporal (slices for different times).  As long as GMT knows it is time we can ensure that the 3rd dimension variable is called time and the units can get similar treatment as x,y, i.e., `units = "seconds since 1970-01-01 00:00:00"`.  If not time then it is called "z" and no info on units is possible.. For the 3-D data value I currently give the data the name "cube" (since I cannot call it "z", obviously) and the data unit is actually given a separate variables in the **GMT_CUBE** header.  But I have no way to set the data unit when creating cubes.

Bottom line is that GMT has no support for standard metadata in grids and cubes: What we refer to as units (header->z_unit; -Dxname sets x-unit name) is actually the name attribute.  The unit is not affected (except for geographic, auto-set things). And we have no ability to name the grid or cube data variable.

So this is my wishlist for grids and cubes when writing creating things from raw data (e.g., from surface, greenspline, sphinterpolate, xyz2grd, ...):

1. Ability to name the grid or cube data variable [z or cube].
2. Ability to specify the grid or cube data unit [not set].
3. Ability to set the unit of the x and y dimensions (if Cartesian) [not set].

What mechanisms can we offer? Well, it is tricky, since options like **grdedit -D** will change values in our grid header but that is limited to the name attribute (remember, no unit attribute even though variable is badly named as x_units).  We could of course add a few new variables in the "hidden" grid header section and then we can access these in gmt_nc.c and do the right thing.  So that would be my preferred route.
How should the user specify this information? One way is to add a few more modifiers to the **-D** option in **grdedit**, say **+ux**|**y**|**z**|**v**_unit_, where this would set the variable unit properly (in netCDF grids). This means surface or greenspline users (all but xyz2grd) will need to use a separate grdedit step, plus grdedit would need an update to handle cubes.  Another mechanism might be via a new lower-case GMT common option.  These are unused so far: **-k**, **-m**, **-u**, **-y**, **-z**.  Could declare that **-m** sets meta-data for grids and offer things like **-m+x**_km_**+y**_km_**+v**_Pa_**+V**_pressure_.  Because GMT does not care about these settings in the files I think our first step should be modest and just let **grdedit** have the ability to properly set the meta data we want.  This would greatly help people who publishes data sets as their data will be more self-documenting.  Hence this proposal:

Make the following changes to **grdedit** and its **-D** option (which is called by **xyz2grd** as well):

1. Let the **-D** option be allowed if file argument is _either_ a grid or cube.
2. Add a new modifier (**+v**_varname_) for setting the data value variable name [z or cube].
3. Add another (**+u**) setting the units for the 2-3 dimensions (this is for Cartesian axes). These would be **+ux**|**y**|**z**|**v**_unit_ and repeatable (i.e., **-D+x**_wavelength_**+ux**_km_**+y**_frequency_**+uy**_Hz_**+z**_depth_**+uz**_km_**+v**_temperature_**+uv**_degrees_. For grids we will allow backwards compatibility with **+z** as it is now (sets attribute name for the grid data).
4. Clean up documentation that confuses attribute name and attribute unit (we only set attribute name via our options).

I would like some feedback from @GenericMappingTools/core on this issue before taking any further action.

PaulWessel commented:
Pinging @remkos to have a look if he has time.

remkos commented:
I was looking at your proposal for units. As far as I remember (but I may have done that in other programs), if one adds "[m]" to the x-name then "m" is used as the unit. This should also hold for y-name, z-name and any other specified units, of course.

But yes, this would be rather "hidden" way of doing things and probably a more direct way of specifying the units, as you suggested above would be better. Or maybe support both.

Indeed, there was no way to specify the variable name. It was always x and y, unless the projection was geographical and lon and lat where used. This is not required by COARDS, but I have seen applications react to it appropriately, even without the units like degrees_east, degrees_north or the appropriate standard_name. Still, what is required by COARDS is that the dimensions and the coordinate axes are called the same. So the dimensions and the coordinate arrays are called lon and lat, consistently.

Now checking the code:
- Yes, gmtnc_put_units does split off the [units] part from the name_units variable which can be either x_units or y_units. So if in the present form you would specify "x-unit" as "temperature [K]" it would write long_name = "temperature" and units = "K".
- There are also some provisions for time axes. But it does not look like that I made any provisions to put the time references in correctly, e.g. "seconds since 2000-01-01 00:00:00".

PaulWessel commented:
It is great that name [unit] is parsed like that - I guess I thought it was a special thing for [degrees_longitude] etc. and not general.  So we can improve the docs on that for sure.

PaulWessel commented:
Since (re-)discovering the ability to specify the name of a dimension or data and add the unit in brackets, the proposal above has been simplified, and implemented.  We now have, via the **-D**_info_ setting, these new things:

1. **+d**_dname_ sets the name and unit of the data value (formerly just **+z**_name_ for grids - this is still backwards compatible) [z]
2. **+z**_name_ sets the 3rd dimension name and unit for cubes [z or time]
3. **+v**_varname_ names the **netCDF** data variable [z or cube]

With that I think we are good.  I updated _gspline_4.sh_ to use **-D** (which I added to greenspline) via the command

`gmt greenspline -R12/32/0/6/5/10 -I0.25 -G3D_cube.grdTable_5_23.txt -Sr0.85 -Z5 -D+x"x-distance [km]"+y"y-distance [km]"+z"z-distance [km]"+d"Uranium Oxide [%]"+vuoxide"`

and now `ncdump -h` gives

```
variables:
	double x(x) ;
		x:long_name = "x-distance" ;
		x:units = "km" ;
		x:actual_range = 12., 32. ;
	double y(y) ;
		y:long_name = "y-distance" ;
		y:units = "km" ;
		y:actual_range = 0., 6. ;
	double z(z) ;
		z:long_name = "z-distance" ;
		z:units = "km" ;
		z:actual_range = 5., 10. ;
	float uoxide(z, y, x) ;
		uoxide:long_name = "Uranium Oxide" ;
		uoxide:units = "%" ;
		uoxide:_FillValue = NaNf ;
		uoxide:actual_range = 0., 23.2000007629395 ;
```

So, the **-D**_info_ was added to both **greenspline** and **grdinterpolate**.  Since **greenspline** already used **-D** I moved that option to **-Z** but we handle this with backwards compatibility.  All tests work fine - I think this PR is now good to be approved and merged.

joa-quim reviewed (approved):
Tests pass for me too.
